src/session.ts

```diff
@@ -1,29 +1,77 @@
 import { Context } from './context'
 import { Middleware } from './types'
 
-export function session<SessionData extends object>({
-  ttl = Infinity,
-  store = new Map<string, { expires: number; session: SessionData }>(),
-  getSessionKey = (ctx: Context) =>
-    ctx.from && ctx.chat && `${ctx.from.id}:${ctx.chat.id}`,
-} = {}): Middleware.ExtFn<Context, { session?: SessionData }> {
-  const ttlMs = ttl * 1000
+export type StorageKeyFn<T extends Context> = (
+  ctx: T
+) => Promise<string | undefined>
 
+export interface Storage<T> {
+  getItem: (name: string) => Promise<T | undefined>
+  setItem: (name: string, value: T) => Promise<void>
+  deleteItem: (name: string) => Promise<void>
+}
+
+export class MemorySessionStorage<T> implements Storage<T> {
+  private readonly ttl: number
+  private readonly store = new Map<string, { session: T; expires: number }>()
+
+  constructor(ttl = Infinity) {
+    this.ttl = ttl * 1000
+  }
+
+  async getItem(name: string): Promise<T | undefined> {
+    const entry = this.store.get(name)
+    if (entry == null) {
+      return undefined
+    } else if (entry.expires < Date.now()) {
+      await this.deleteItem(name)
+      return undefined
+    }
+    return entry.session
+  }
+
+  async setItem(name: string, value: T): Promise<void> {
+    const now = Date.now()
+    this.store.set(name, { session: value, expires: now + this.ttl })
+  }
+
+  async deleteItem(name: string): Promise<void> {
+    this.store.delete(name)
+  }
+}
+
+/**
+ * session middleware
+ *
+ * more 3rd-party session middlware
+ *
+ * see https://npm.im/search?q=telegraf-session
+ */
+export function session<SessionData extends object>(
+  storage: Storage<SessionData> = new MemorySessionStorage<SessionData>(),
+  makeKey: StorageKeyFn<Context> = makeDefaultKey
```

Comment:
I think these should be swapped
```suggestion
  makeKey: StorageKeyFn<Context> = makeDefaultKey,
  storage: Storage<SessionData> = new MemorySessionStorage<SessionData>()
```

or even better, put into an object.